File: app/src/main/java/com/diamond/appcliente/actividades/GestionarServicioActivity.java
package com.diamond.appcliente.actividades;

import android.os.Bundle;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.diamond.appcliente.R;
import com.diamond.appcliente.adapters.ServicioAdapter;
import com.diamond.appcliente.dto.servicio.ServicioDto;
import com.diamond.appcliente.dto.servicio.ServicioRequest;
import com.diamond.appcliente.viewmodel.GestionarServicioViewModel;

import java.util.List;

public class GestionarServicioActivity extends AppCompatActivity {

    private GestionarServicioViewModel viewModel;
    private RecyclerView recyclerView;
    private Button btnAgregarServicio;
    private ServicioAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_gestionar_servicio);

        recyclerView = findViewById(R.id.recyclerViewServicios);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 2)); // 2 columnas

        btnAgregarServicio = findViewById(R.id.btnAgregarServicio);

        // Inicializa el ViewModel
        viewModel = new ViewModelProvider(this).get(GestionarServicioViewModel.class);

        // Observa el LiveData para actualizar la UI cuando los servicios cambien
        // Cargar los servicios
        viewModel.obtenerServicios(this, new GestionarServicioViewModel.ServicioCallback() {
            @Override
            public void onSuccess(List<ServicioDto> servicios) {
                // Si la lista se actualiza, actualizamos el adaptador
                if (adapter == null) {
                    adapter = new ServicioAdapter(servicios, new ServicioAdapter.OnServicioClickListener() {
                        @Override
                        public void onAviso(ServicioDto servicio, String imagenUrl) {

                        }
                    });
                    recyclerView.setAdapter(adapter);
                } else {
                    // Si ya existe el adaptador, solo actualizamos la lista
                    adapter.notifyDataSetChanged();
                }
            }

            @Override
            public void onError(String mensaje) {
                // Si ocurre un error, muestra un mensaje en el toast
                Toast.makeText(GestionarServicioActivity.this, "Error: " + mensaje, Toast.LENGTH_SHORT).show();
            }
        });
    }


    private void crearNuevoServicio(String nombre, double precio, String descripcion, int tipoServicioId) {
        ServicioRequest request = new ServicioRequest(nombre, precio, descripcion, tipoServicioId);
        viewModel.crearServicio(this, request, new GestionarServicioViewModel.ServicioOperacionCallback() {
            @Override
            public void onSuccess(String mensaje) {
                Toast.makeText(GestionarServicioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                // Refrescar la lista de servicios después de crear uno nuevo
                viewModel.obtenerServicios(GestionarServicioActivity.this, new GestionarServicioViewModel.ServicioCallback() {
                    @Override
                    public void onSuccess(List<ServicioDto> servicios) {
                        adapter.notifyDataSetChanged();
                    }

                    @Override
                    public void onError(String mensaje) {
                        Toast.makeText(GestionarServicioActivity.this, "Error: " + mensaje, Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onError(String mensaje) {
                Toast.makeText(GestionarServicioActivity.this, "Error: " + mensaje, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void eliminarServicio(int id) {
        viewModel.eliminarServicio(this, id, new GestionarServicioViewModel.ServicioOperacionCallback() {
            @Override
            public void onSuccess(String mensaje) {
                Toast.makeText(GestionarServicioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                // Refrescar la lista después de eliminar un servicio
                viewModel.obtenerServicios(GestionarServicioActivity.this, new GestionarServicioViewModel.ServicioCallback() {
                    @Override
                    public void onSuccess(List<ServicioDto> servicios) {
                        adapter.notifyDataSetChanged();
                    }

                    @Override
                    public void onError(String mensaje) {
                        Toast.makeText(GestionarServicioActivity.this, "Error: " + mensaje, Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onError(String mensaje) {
                Toast.makeText(GestionarServicioActivity.this, "Error: " + mensaje, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/diamond/appcliente/actividades/GestionarServicioActivity.kt
package com.diamond.appcliente.actividades

import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.activity.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.diamond.appcliente.R
import com.diamond.appcliente.adapters.ServicioAdapter
import com.shared.models.ui.state.UiState
import com.diamond.appcliente.viewmodel.GestionarServicioViewModel
import com.diamond.barbershop.shared.dto.servicio.ServicioDto
import com.diamond.barbershop.shared.dto.servicio.ServicioRequest
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class GestionarServicioActivity : AuthActivity() {

    private val viewModel: GestionarServicioViewModel by viewModels()
    private lateinit var recyclerView: RecyclerView
    private lateinit var btnAgregarServicio: Button
    private var adapter: ServicioAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_gestionar_servicio)

        recyclerView = findViewById(R.id.recyclerViewServicios)
        recyclerView.layoutManager = GridLayoutManager(this, 2)
        btnAgregarServicio = findViewById(R.id.btnAgregarServicio)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.servicios.collect { state ->
                        when (state) {
                            is UiState.Success -> {
                                val servicios = state.data
                                if (adapter == null) {
                                    adapter = ServicioAdapter(servicios, object : ServicioAdapter.OnServicioClickListener {
                                        override fun onAviso(servicio: ServicioDto, imagenUrl: String?) {}
                                    })
                                    recyclerView.adapter = adapter
                                } else {
                                    adapter!!.notifyDataSetChanged()
                                }
                            }
                            is UiState.Error -> Toast.makeText(
                                this@GestionarServicioActivity, state.message, Toast.LENGTH_SHORT
                            ).show()
                            else -> {}
                        }
                    }
                }
                launch {
                    viewModel.mensaje.collect { msg ->
                        Toast.makeText(this@GestionarServicioActivity, msg, Toast.LENGTH_SHORT).show()
                    }
                }
                launch {
                    viewModel.crudError.collect { msg ->
                        Toast.makeText(this@GestionarServicioActivity, "Error: $msg", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }

        viewModel.cargarServicios()
    }

    private fun crearNuevoServicio(nombre: String, precio: Double, descripcion: String, tipoServicioId: Int) {
        viewModel.crearServicio(ServicioRequest(nombre, precio, descripcion, tipoServicioId))
    }

    private fun eliminarServicio(id: Int) {
        viewModel.eliminarServicio(id)
    }
}
